cmd/snap-confine/mount-support.c

```diff
@@ -248,6 +248,206 @@ struct sc_mount_config {
+
+	// Will create folders/links as 0:0
+	sc_identity old = sc_set_effective_identity(sc_root_group_identity());
+
+	char full_path[PATH_MAX];
+	// After we construct each entry's full path, we'll need to obtain the
+	// entry's absolute path in the new rootfs, that is with the `scratch_dir`
+	// prefix removed (the path_in_rootfs variable below). We'll do this by
+	// computing the length of the `scratch_dir` prefix now and then using it
+	// as the offset in `full_path` where the '/' of the confined silesystem is
+	// located.
+	const size_t scratch_dir_length = strlen(scratch_dir);
+
+	while (true) {
+		errno = 0;
+		struct dirent *ent = readdir(rootfs);
+		if (ent == NULL) break;
+
+		if (sc_streq(ent->d_name, ".") || sc_streq(ent->d_name, "..")) {
+			continue;
+		}
+
+		sc_must_snprintf(full_path, sizeof(full_path), "%s/%s",
+		                 scratch_dir, ent->d_name);
+		if (ent->d_type == DT_DIR) {
+			if (mkdir(full_path, 0755) < 0) {
+				die("cannot create directory \"%s\"", full_path);
+			}
+
+			// If the directory is listed in root_mounts skip it,
+			// as it will be created and mounted in
+			// sc_bootstrap_mount_namespace() later.
+			bool skip_dir = false;
+			const char *path_in_rootfs = full_path + scratch_dir_length;
+			for (const struct sc_mount *mnt = root_mounts;
+			     mnt->path != NULL; mnt++) {
+				if (sc_streq(path_in_rootfs, mnt->path) ||
+				    sc_streq(path_in_rootfs, mnt->altpath)) {
+					skip_dir = true;
+					break;
+				}
+			}
+			if (skip_dir) {
+				continue;
+			}
+			// Also skip the /snap directory, as we'll mount it later
+			if (sc_streq(path_in_rootfs, "/snap")) {
+				skip_dir = true;
+			}
+
+			char src_path[PATH_MAX];
+			sc_must_snprintf(src_path, sizeof(src_path), "%s/%s",
+					 rootfs_dir, ent->d_name);
+			sc_do_mount(src_path, full_path, NULL, MS_REC | MS_BIND, NULL);
+			sc_do_mount("none", full_path, NULL, MS_REC | MS_SLAVE, NULL);
+		} else if (ent->d_type == DT_LNK) {
+			char link_target[PATH_MAX + 1];
+			ssize_t len = readlinkat(rootfs_fd, ent->d_name,
+			                         link_target, sizeof(link_target) - 1);
+			if (len < 0) {
+				die("cannot read symbolic link \"%s/%s\"",
+				    rootfs_dir, ent->d_name);
+			}
+			// make sure the string is null terminated
+			link_target[len] = '\0';
+
+			// Both relative and absolute links will work out of the box, since
+			// we are going to do a pivot_root to scratch_dir.
+			if (symlink(link_target, full_path) < 0) {
+				die("cannot create symbolic link \"%s\"", full_path);
+			}
+		} else if (ent->d_type == DT_REG) {
+			// Create an empty file which can be used as a mount point
+			int fd = open(full_path, O_CREAT | O_TRUNC, 0644);
+			if (fd < 0) {
+				die("cannot create mount point for file \"%s\"", full_path);
+			}
+			close(fd);
+			char src_path[PATH_MAX];
+			sc_must_snprintf(src_path, sizeof(src_path), "%s/%s",
+					 rootfs_dir, ent->d_name);
+			sc_do_mount(src_path, full_path, NULL, MS_BIND, NULL);
+		}
```

Comment:
Should we have an error or a warning when some other type of entries exist and are ignored? This should never happen. But it would be nice to know if it did.

Reply:
I agree - particularly if we are sure this should never happen, it would be good to capture this invariant with a die() or similar.

Reply:
Good points! I added a `die` case should we encounter a type we do not expect

Reply:
Looks fine. Resolved.